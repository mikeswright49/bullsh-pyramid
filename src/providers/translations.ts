export const TRANSLATIONS = {
    en: {
        'game.board.current.row': 'Current Row: {{ activeRow }}',
        'game.board.current.card': 'Current card',
        'game.display.stage.memorization.text': 'You have 30s to remember your cards!',
        'game.display.stage.flipping.text': 'The host is flipping a card',
        'game.display.stage.declaration.text': 'Let the world know if you have it',
        'game.display.stage.assign.text': 'Players are beginning to lie left right and center',
        'game.display.stage.bullshit.text': 'People should know who you think was lieing',
        'game.display.stage.reveal.text': 'We all now know the truth',
        'game.display.stage.memory.text': 'Well which cards did you have?',
        'game.display.stage.complete.text': 'Well time for a new game',
    },
};
