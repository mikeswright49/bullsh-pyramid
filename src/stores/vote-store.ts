import { BaseStore } from './base-store';
import { shortId } from 'src/utilities/shortid';
import { Vote } from 'types/vote';
import { Card } from 'types/card';
import { PlayerStore } from './player-store';
import { Player } from 'types/player';

export class VoteStore extends BaseStore {
    public static async createVote({
        amount,
        playerId,
        targetId,
        card,
    }: {
        amount: number;
        playerId: string;
        targetId: string;
        card: Card;
    }): Promise<string> {
        const voteId = shortId();
        try {
            await VoteStore.database.ref(`votes/${voteId}`).set({ amount, card });
            await VoteStore.database.ref(`votes/${voteId}/playerref`).push(playerId);
            await VoteStore.database.ref(`votes/${voteId}/targetref`).push(targetId);
            return voteId;
        } catch (error) {
            console.error(error);
        }
    }

    public static async addVote(gameId: string, voteId: string) {
        try {
            await VoteStore.database.ref(`/games/${gameId}/votes`).push(voteId);
        } catch (error) {
            console.error(error);
        }
    }

    public static async getTarget(voteId: string): Promise<Player> {
        const ref = VoteStore.database.ref(`votes/${voteId}/targetref`);
        return VoteStore.getPlayerRef(ref);
    }

    public static async getPlayer(voteId: string): Promise<Player> {
        const ref = VoteStore.database.ref(`votes/${voteId}/playerref`);
        return VoteStore.getPlayerRef(ref);
    }

    public static async removeVotes(gameId: string) {
        try {
            const votes = await VoteStore.database.ref(`/games/${gameId}/votes`).once('value');
            await Promise.all(
                votes.val().map((childId: string) => {
                    return votes.child(childId).ref.remove();
                })
            );
        } catch (error) {
            console.error(error);
        }
    }

    public static subscribeToVotes(gameId: string, callback: (vote: Vote) => void) {
        if (!gameId) {
            return;
        }
        const game = VoteStore.database.ref(`games/${gameId}/votes`);
        const votes = VoteStore.database.ref(`votes`);
        game.on('child_added', (child) => {
            const voteId = child.val();
            votes.child(voteId).on('value', async (val) => {
                const value: Vote = val.val();
                value.id = voteId;
                value.player = await VoteStore.getPlayer(voteId);
                value.target = await VoteStore.getTarget(voteId);
                callback(value as Vote);
            });
        });
    }

    private static async getPlayerRef(ref: firebase.database.Reference): Promise<Player> {
        const playerId: string = (await ref.once('child_added')).val();
        return PlayerStore.getPlayer(playerId);
    }
}
