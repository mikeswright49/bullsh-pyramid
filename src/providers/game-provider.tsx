import React, { useEffect, useState } from 'react';
import { GameState } from 'types/game-state';
import { GameStore } from 'src/stores/game-store';
import { GameStage } from 'src/enums/game-stage';

export const DEFAULT_GAME_STATE: GameState = {
    id: '',
    gameStage: GameStage.Initiation,
    activeRow: 0,
    activeIndex: 0,
    cardCount: 4,
    flipDelay: 5000,
    players: [],
    tiers: [],
};
export const GameContext = React.createContext<GameState>(DEFAULT_GAME_STATE);

export function GameProvider(props: React.PropsWithChildren<{ gameId: string }>) {
    const { gameId } = props;
    const [gameState, setGameState] = useState<GameState>(DEFAULT_GAME_STATE);

    useEffect(() => {
        if (!gameId) {
            return;
        }

        GameStore.subscribeToGame(gameId, (updatedGameState) => {
            setGameState(updatedGameState);
        });

        return () => GameStore.unsubscribeToGame();
    }, [gameId]);

    return <GameContext.Provider value={gameState}>{props.children}</GameContext.Provider>;
}
