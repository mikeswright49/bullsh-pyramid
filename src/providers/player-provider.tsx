import React from 'react';
import { useState, useEffect } from 'react';
import { Player } from 'types/player';
import { PlayerStore } from 'src/stores/player-store';
export const PlayerContext = React.createContext<Player>(null);

export function PlayerProvider(props: React.PropsWithChildren<{ playerId: string }>) {
    const [player, setPlayer] = useState<Player>();

    useEffect(() => {
        if (!props.playerId) {
            return;
        }
        PlayerStore.subscribeToPlayer(props.playerId, (updatedPlayer) => {
            setPlayer(updatedPlayer);
        });

        return () => PlayerStore.unsubscribeToPlayer();
    }, [props.playerId]);

    return <PlayerContext.Provider value={player}>{props.children}</PlayerContext.Provider>;
}
