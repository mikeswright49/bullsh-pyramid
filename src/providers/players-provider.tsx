import React from 'react';
import { useState, useEffect } from 'react';
import { Player } from 'types/player';
import { GameStore } from 'src/stores/game-store';

export const PlayersContext = React.createContext<Player[]>([]);

export function PlayersProvider(props: React.PropsWithChildren<{ gameId: string }>) {
    const [players, setPlayers] = useState({});
    useEffect(() => {
        GameStore.subscribeToPlayers(props.gameId, (player: Player) => {
            players[player.id] = player;
            setPlayers({ ...players });
        });
    }, [props.gameId]);
    return (
        <PlayersContext.Provider value={Object.values(players)}>
            {props.children}
        </PlayersContext.Provider>
    );
}
