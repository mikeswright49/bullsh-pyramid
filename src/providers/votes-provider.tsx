import React from 'react';
import { useState, useEffect } from 'react';
import { Vote } from 'types/vote';
import { VoteStore } from 'src/stores/vote-store';
export const VotesContext = React.createContext<Vote[]>([]);

export function VotesProvider(props: React.PropsWithChildren<{ gameId: string }>) {
    const [votes, setVotes] = useState([]);
    useEffect(() => {
        if (!votes) {
            return;
        }

        VoteStore.subscribeToVotes(props.gameId, (vote: Vote) => {
            votes[vote.id] = vote;
            setVotes({ ...votes });
        });
    }, [props.gameId]);

    return (
        <VotesContext.Provider value={Object.values(votes)}>{props.children}</VotesContext.Provider>
    );
}
