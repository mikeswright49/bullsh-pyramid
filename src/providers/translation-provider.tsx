import React, { useContext } from 'react';
import { useEffect, useState } from 'react';
import { GameContext } from './game-provider';
import { TRANSLATIONS } from '../translations/translations';

const INTERPOLATE_REGEX = /{{.*?}}/gi;
const INTERPOLATE_PROPERTY_REGEX = /{{(.*)?}}/i;
const INTERPOLATE_CAPTURE_COUNT = 2;

export type Translations = {
    translations: { [key: string]: string };
    translate: (key: string, variables?: { [key: string]: string | number | boolean }) => string;
};

export const TranslationContext = React.createContext<Translations>({
    translate: (a) => a,
    translations: {},
});

export function TranslationProvider(props: React.PropsWithChildren<unknown>) {
    const { language } = useContext(GameContext);
    const [translations, setTranslations] = useState({});
    useEffect(() => {
        setTranslations(TRANSLATIONS[language || 'en']);
    }, [language]);

    /**
     * Returns the nested property value (userMetadata.giftcard.bacon)
     * @param source property value
     */
    function getProperty(source: string[], parent: unknown): string {
        if (!source || !parent) {
            return;
        }
        if (source.length > 1) {
            const property: string = source.shift();
            return this.getProperty(source, parent[property.trim()]);
        }
        return parent[source[0].trim()];
    }

    /**
     * Evaluates the {{}} interpolated strings
     * @param source Source content to be parsed
     * @param data Data object to read values off of
     */
    function evaluateInterpolate(
        source: string,
        data: { [key: string]: string | number | boolean }
    ): string {
        if (!data) {
            return source;
        }

        const matches = source.match(INTERPOLATE_REGEX) || [];
        matches.forEach((match: string) => {
            const properties = INTERPOLATE_PROPERTY_REGEX.exec(match);
            if (properties && properties.length === INTERPOLATE_CAPTURE_COUNT) {
                const property = getProperty(properties[1].split('.'), data);
                source = source.replace(match, property || '');
            }
        });
        return source;
    }

    function translate(key: string, variables?: { [key: string]: string | number | boolean }) {
        if (!translations) {
            return '';
        }
        return evaluateInterpolate(translations[key], variables);
    }

    return (
        <TranslationContext.Provider value={{ translations, translate }}>
            {props.children}
        </TranslationContext.Provider>
    );
}
