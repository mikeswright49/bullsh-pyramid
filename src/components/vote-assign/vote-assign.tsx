import React, { useContext, useRef, useState } from 'react';
import { PlayersContext } from 'src/providers/players-provider';
import { PlayerContext } from 'src/providers/player-provider';
import { VoteStore } from 'src/stores/vote-store';
import { GameContext } from 'src/providers/game-provider';
import { Card as CardType } from 'types/card';
import { Card } from '../card/card';
export interface VoteAssignProps {
    card: CardType;
}

export function VoteAssign({ card }: VoteAssignProps) {
    const players = useContext(PlayersContext);
    const player = useContext(PlayerContext);
    const gameState = useContext(GameContext);

    const [errorMessage, setErrorMessage] = useState('');
    const voteInputs = useRef([]);
    const [submitted, setSubmitted] = useState(false);

    const otherPlayers = players.filter((otherPlayer) => otherPlayer.id !== player.id);
    const count = gameState.activeRow + 1;

    async function createVotes(event: React.FormEvent): Promise<void> {
        event.preventDefault();
        let totalVotes = 0;
        const assignedPlayers = [];

        voteInputs.current.forEach((input, index) => {
            const assigned = parseInt(input.value, 10);
            totalVotes += assigned;
            if (assigned > 0) {
                assignedPlayers.push({
                    playerId: player.id,
                    targetId: otherPlayers[index].id,
                    amount: assigned,
                    card,
                });
            }
        });

        if (totalVotes !== count) {
            setErrorMessage('You must assign votes out');
            return;
        }

        await Promise.all(
            assignedPlayers.map(async (assignee) => {
                const voteId = await VoteStore.createVote(assignee);
                await VoteStore.addVote(gameState.id, voteId);
            })
        );

        setSubmitted(true);
    }

    return (
        <div data-testid="vote-assign">
            {!submitted ? (
                <>
                    <h4>Assign {count} &quot;points&quot; out</h4>
                    <form
                        onSubmit={createVotes}
                        data-testid="vote-assign-form"
                        style={{
                            display: 'flex',
                            alignItems: 'center',
                        }}
                    >
                        <Card card={card} />
                        <div>
                            <div className="row">
                                <h4 className="col-8-sm">Player name</h4>
                                <h4 className="col-3-sm">Points</h4>
                            </div>
                            {otherPlayers.map((player, index) => {
                                return (
                                    <div key={player.id} className="row">
                                        <label
                                            className="col-8-sm"
                                            htmlFor={`player-${player.id}-amount`}
                                        >
                                            {player.name}
                                        </label>
                                        <input
                                            className="col-3-sm"
                                            type="number"
                                            defaultValue={0}
                                            max={10}
                                            ref={(ref) => (voteInputs.current[index] = ref)}
                                            data-testid={`player-${player.id}-amount`}
                                            name={`player-${player.id}-amount`}
                                        ></input>
                                    </div>
                                );
                            })}
                            {errorMessage && <div>{errorMessage}</div>}
                            <button type="submit">All done</button>
                        </div>
                    </form>
                </>
            ) : (
                <h5>That was pretty easy</h5>
            )}
        </div>
    );
}
