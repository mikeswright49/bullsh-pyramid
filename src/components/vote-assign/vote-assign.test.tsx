import React from 'react';
import { VoteAssignProps, VoteAssign } from './vote-assign';
import { PlayersContext } from 'src/providers/players-provider';
import { GameContext } from 'src/providers/game-provider';
import { PlayerContext } from 'src/providers/player-provider';
import { render, fireEvent, act } from '@testing-library/react';
import { MOCK_GAME_STATE } from 'src/__tests__/data/mock-game-state';
import { MOCK_PLAYER } from 'src/__tests__/data/mock-player';
import { MOCK_PLAYERS } from 'src/__tests__/data/mock-players';
import { VoteStore } from 'src/stores/vote-store';

jest.mock('src/stores/vote-store');

describe('VoteAssign', () => {
    let mockProps: VoteAssignProps;
    let container: HTMLElement;

    function renderComponent() {
        container = render(
            <GameContext.Provider value={MOCK_GAME_STATE}>
                <PlayersContext.Provider value={MOCK_PLAYERS}>
                    <PlayerContext.Provider value={MOCK_PLAYER}>
                        <VoteAssign {...mockProps} />
                    </PlayerContext.Provider>
                </PlayersContext.Provider>
            </GameContext.Provider>
        ).container;
    }

    beforeEach(() => {
        mockProps = {
            card: { value: 1, suit: 0, hidden: false },
        };

        ((VoteStore.createVote as unknown) as jest.Mock).mockReturnValue(
            Promise.resolve('a-vote-id')
        );
        ((VoteStore.addVote as unknown) as jest.Mock).mockReturnValue(Promise.resolve());
    });

    it('should be able to render the component', () => {
        renderComponent();
        expect(container.querySelector('[data-testid="vote-assign"]')).toMatchSnapshot();
    });
    it('should be able to create the votes for a given set up', async () => {
        renderComponent();
        fireEvent.change(container.querySelector('[data-testid="player-ffy-amount"]'), {
            target: { value: '2' },
        });

        await act(async () =>
            fireEvent.submit(container.querySelector('[data-testid="vote-assign-form"]'))
        );

        expect(container.querySelector('[data-testid="vote-assign"]')).toMatchSnapshot();
        expect(VoteStore.createVote).toHaveBeenCalledWith(
            expect.objectContaining({ amount: 2, playerId: 'nez', targetId: 'ffy' })
        );
        expect(VoteStore.addVote).toHaveBeenCalledWith('mwif', 'a-vote-id');
    });
});
