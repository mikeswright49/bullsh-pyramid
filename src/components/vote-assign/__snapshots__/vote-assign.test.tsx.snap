// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`VoteAssign should be able to create the votes for a given set up 1`] = `
<div
  data-testid="vote-assign"
>
  <h5>
    That was pretty easy
  </h5>
</div>
`;

exports[`VoteAssign should be able to render the component 1`] = `
<div
  data-testid="vote-assign"
>
  <h4>
    Assign 
    2
     "points" out
  </h4>
  <form
    data-testid="vote-assign-form"
    style="display: flex; align-items: center;"
  >
    <div
      class="flex-center-center card"
      data-testid="card-0-1"
      style="background-position: -79px -0px;"
    />
    <div>
      <div
        class="row"
      >
        <h4
          class="col-8-sm"
        >
          Player name
        </h4>
        <h4
          class="col-3-sm"
        >
          Points
        </h4>
      </div>
      <div
        class="row"
      >
        <label
          class="col-8-sm"
          for="player-ffy-amount"
        >
          joe
        </label>
        <input
          class="col-3-sm"
          data-testid="player-ffy-amount"
          max="10"
          name="player-ffy-amount"
          type="number"
          value="0"
        />
      </div>
      
      <button
        type="submit"
      >
        All done
      </button>
    </div>
  </form>
</div>
`;
